### PR TITLE
Add MAML meta-train prototype code

### PR DESCRIPTION
**Pull request by Clyde21c**
## PR Description

This PR adds MAML meta-train code and modifies some formatting aligning with RL^2.

## Checklist

- [x] Assignee's formatting
- [x] Assignee's linting
- [x] Reviewer's code runnability

### CODE REVIEW

**Comment by Clyde21c** [comment]
> @Clyde21c Did you check formatting and linting? If you checked them, please mark them on the checklist above.

Confirmed it. marked the checklist

File: maml/algorithm/meta_learner.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    ):
+
+        self.env = env
+        self.env_name = env_name
+        self.agent = agent
+        self.train_tasks = train_tasks
+        self.test_tasks = test_tasks
+
+        self.train_iters = config["train_iters"]
+        self.num_sample_tasks = config["num_sample_tasks"]
```
Comment (dongminlee94) [inline maml/algorithm/meta_learner.py:43]: RL^2, PEARL과 동일하게 `train_tasks`로 바꾸면 어떨까요? 혹시 `num_sample_tasks`로 바꾸신 이유가 있을까요?

File: maml/algorithm/meta_learner.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            action_dim=action_dim,
+            max_size=self.train_samples * self.num_sample_tasks,
+            device=device,
+        )
+        self.inner_buffer = Buffer(
+            observ_dim=observ_dim,
+            action_dim=action_dim,
+            max_size=self.train_samples,
+            device=device,
+        )
```
Comment (dongminlee94) [inline maml/algorithm/meta_learner.py:69]: buffer를 두개 만드는 이유는 무엇인가요? 그리고 buffer를 두개 쓰는 reference code가 있다면, 링크 걸어주시면 감사하겠습니다.

File: maml/algorithm/meta_learner.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        self.meta_optimizer = torch.optim.Adam(
+            list(self.agent.policy.parameters()),
+            lr=config["learning_rate"],
+        )
+        self.inner_optimizer = torch.optim.SGD(
+            list(self.agent.policy.parameters()),
+            lr=config["inner_learning_rate"],
+        )
+
+        if file_name is None:
```
Comment (dongminlee94) [inline maml/algorithm/meta_learner.py:80]: 아마 제 코드에도 바꿔야할 것 같은데, `is None`은 지워도 될 것 같습니다. file_name이 none이면 false로 인식해서 if문에 안들어가서요.

reviewer의 시선으로 보니 이런것도 보이는군요 ㅎㅎ

File: maml/algorithm/meta_learner.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        )
+
+    def meta_train(self):
+        """MAML meta-training"""
+        for iteration in range(self.train_iters):
+
+            print("=============== Iteration {} ===============".format(iteration))
+            # Sample tasks randomly from train tasks distribution.
+            for _ in range(self.num_sample_tasks):
+                index = np.random.randint(len(self.train_tasks))
```
Comment (dongminlee94) [inline maml/algorithm/meta_learner.py:98]: 이 부분 저번에 RL^2와 비교할 때 한번 논의가 되었던 것 같은데, 그 때 어떻게 결정하기로 했었죠? RL^2와 MAML 동일하게 task를 뽑는 형식으로 갈까요? 아니면 서로 다르게 갈까요?

File: maml/algorithm/meta_learner.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                        index + 1, len(self.train_tasks)
+                    )
+                )
+
+                # Branch policy and optimizer
+                with higher.innerloop_ctx(
+                    self.agent.policy,
+                    self.inner_optimizer,
+                    copy_initial_weights=False,
+                ) as (inner_policy, branch_optimizer):
```
Comment (dongminlee94) [inline maml/algorithm/meta_learner.py:112]: higher가 그러니까 deepcopy를 하는 것처럼 branch를 나눠서 (다른 메모리를 할당받아서) 작동하는 방식인가요? 이 부분에 대해 어떻게 작동하는지 공유가 되면 좋을 것 같습니다.

File: maml/algorithm/meta_learner.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                    self.meta_buffer.add_trajs(trajs)
+
+            # Meta update
+            print("meta_buffer")
+            batch = self.meta_buffer.get_samples()
+            for _ in range(self.meta_grad_iters):
+                meta_loss = self.agent.compute_losses(self.agent.policy, batch)
+                self.meta_optimizer.zero_grad()
+                meta_loss.backward()
+                self.meta_optimizer.step()
```
Comment (dongminlee94) [inline maml/algorithm/meta_learner.py:144]: PPO에서 사용하는 mini batch 방식은 MAML 알고리즘에는 포함이 안되나요?

File: maml/algorithm/networks.py [hunk 7]
```diff
@@ -4,8 +4,8 @@
 
 import numpy as np
 import torch
-import torch.nn as nn
 import torch.nn.functional as F
+from torch import nn
```
Comment (dongminlee94) [inline maml/algorithm/networks.py:8]: 호기심에 여쭤봅니다. 이렇게 바꾸신 이유가 있을까요?

File: maml/algorithm/ppo.py [hunk 8]
```diff
@@ -4,9 +4,9 @@
 
 import torch
 import torch.nn.functional as F
-import torch.optim as optim
+from torch import optim
```
Comment (dongminlee94) [inline maml/algorithm/ppo.py:7]: 호기심에 여쭤봅니다. 이렇게 바꾸신 이유가 있을까요?

File: maml/algorithm/ppo.py [hunk 9]
```diff
@@ -16,95 +16,60 @@ def __init__(  # pylint: disable=too-many-arguments
...
             input_dim=observ_dim,
             output_dim=1,
-            hidden_units=hidden_units,
+            hidden_dims=hidden_dims,
         ).to(device)
-
-        self.optimizer = optim.Adam(
-            list(self.policy.parameters()) + list(self.vf.parameters()),
-            lr=config["vf_lr"],
+        # self.vf.to(device)
```
Comment (dongminlee94) [inline maml/algorithm/ppo.py:40]: 주석 지워야할 것 같습니다.

File: maml/configs/experiment_config.yaml [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+# Experiment configs
+
+# Select either cheetah-dir or cheetah-vel
+env_name: "cheetah-dir"
+
+# Set an experiment name
+exp_name: "exp_1"
+
+# Set a file name
+file_name: "train_tasks_0_0"
```
Comment (dongminlee94) [inline maml/configs/experiment_config.yaml:11]: 초기 세팅에는 null로 바꿔주시기 바랍니다. yaml에서는 null -> None으로 인식됩니다.

Reference: https://stackoverflow.com/questions/30134110/how-can-i-output-blank-value-in-python-yaml-file/37445121

**Review by dongminlee94 — COMMENTED** [review]
리뷰 comment 드립니다.

**Comment by dongminlee94** [comment]
@Clyde21c 추가적으로, higher가 새롭게 들어오게 되서 `requirements.txt` 파일에도 반영해야할 것 같습니다. runnability를 체크하려고 하니까 higher가 없다고 뜨네요. `requirements.txt`에 해당 version 명시하시고 commit하시면, 그때 runnability를 체크하겠습니다.

File: maml/algorithm/meta_learner.py (hunk 2 above)
Comment (Clyde21c) [reply maml/algorithm/meta_learner.py:69]: 각 태스크 별로 inner loop에서 trajs샘플 후 1차 업데이트 후 다시 trajs 샘플링을 하고 그걸 모아둬야 모든 태스크가 끝난 outer loop에서 업데이트를 하는데요, inner loop샘플과 outer loop샘플을 분리하려고 두개 만들었어요

File: maml/algorithm/meta_learner.py (hunk 3 above)
Comment (Clyde21c) [reply maml/algorithm/meta_learner.py:80]: 넵 ㅋㅋ 수정하겠습니다

File: maml/algorithm/meta_learner.py (hunk 1 above)
Comment (Clyde21c) [reply maml/algorithm/meta_learner.py:43]: 아 넵. 아래 코멘트랑 이어지는 내용인데요, maml 원 알고리즘과 레퍼 코드에서 전체 task를 사용하는게 아니라 train task set에서 몇개의 task를 샘플 해서 학습하더라구요. 그래서 몇개의 task를 샘플할지를 정해주는 항목입니다. 이름은 레퍼런스거 그대로 사용했어요

File: maml/algorithm/meta_learner.py (hunk 4 above)
Comment (Clyde21c) [reply maml/algorithm/meta_learner.py:98]: 네 maml은 논문도 그렇고 레퍼도 그렇고 task 샘플링이 있어서 그렇게 갈까해요. RL^2는 한번에 task 전체 다 학습하는걸로 짜지 않으셨나요?? 현실적으로 RL^2가 300개 task면 샘플하는것도 괜찮을것 같네요

File: maml/algorithm/meta_learner.py (hunk 5 above)
Comment (Clyde21c) [reply maml/algorithm/meta_learner.py:112]: 네 브랜치 나눠서 하는걸로 이해하긴 했는데, 좀더 실험을 해볼까 합니다ㅋㅋ 주석으로 설명도 넣어야 할것 같아요. 실험하고나서 공유드릴게요

File: maml/algorithm/meta_learner.py (hunk 6 above)
Comment (Clyde21c) [reply maml/algorithm/meta_learner.py:144]: 네 MAML은 default가 1 step gradient를 통한 adaptation을 목표로 하고 있어서 mini batch는 여러번 gradient step을 전제로 하고 있기 때문에 고민해본 결과 포함 안하는게 맞다고 생각 됐어요. 레퍼에도 그렇게 구현 됐더라구요

File: maml/algorithm/networks.py (hunk 7 above)
Comment (Clyde21c) [reply maml/algorithm/networks.py:8]: 아 이건 린트에서 뭐가 막혀서 그랬던것 같은데 요건 기억이 잘 안나네요 흠 ㅋㅋ

File: maml/algorithm/ppo.py (hunk 8 above)
Comment (Clyde21c) [reply maml/algorithm/ppo.py:7]: 이건 확실하게 제 컴퓨터에선 린트에서 오류떠서 바꿔줬어요

File: maml/configs/experiment_config.yaml (hunk 10 above)
Comment (Clyde21c) [reply maml/configs/experiment_config.yaml:11]: 네 안그래도 최근 커밋에 그렇게 하셨더라구요. 다음 커밋에 반영하겠습니다.

**Comment by Clyde21c** [comment]
@dongminlee94 넵 requirement에 higher 추가하겠습니다. 1차 리뷰 감사합니다 :1st_place_medal:

File: maml/algorithm/meta_learner.py (hunk 2 above)
Comment (dongminlee94) [reply maml/algorithm/meta_learner.py:69]: 원래 참고하셨던 코드에서는 어떻게 하나요? 링크가 있을까요?

File: maml/algorithm/meta_learner.py (hunk 1 above)
Comment (dongminlee94) [reply maml/algorithm/meta_learner.py:43]: 아 넵 좋습니다 이부분은 PEARL도 원래 코드는 `num_sample_tasks`로 되어있는데요. 제가 드리고 싶은 말씀은 통일이 되었으면 해서요. `num_sample_tasks`로 가도 좋구요. 기존에 `train_tasks`도 좋구요. 하나로 통일하면 좋을 것 같습니다.

File: maml/algorithm/meta_learner.py (hunk 4 above)
Comment (dongminlee94) [reply maml/algorithm/meta_learner.py:98]: 음 그러면 300개 중에 어떤 task는 학습을 하지 않을수도 있겠네요. 추후에 MAML처럼 추가 실험 한번더 해보겠습니다. 이부분도 동일하게 가면 좋겠네요.

Comment (dongminlee94) [reply maml/algorithm/meta_learner.py:98]: Done

File: maml/algorithm/meta_learner.py (hunk 5 above)
Comment (dongminlee94) [reply maml/algorithm/meta_learner.py:112]: OK. agreed.

File: maml/algorithm/meta_learner.py (hunk 6 above)
Comment (dongminlee94) [reply maml/algorithm/meta_learner.py:144]: reference 링크가 있을까요? 해당 줄까지 올려주시면 감사하겠습니다.

File: maml/algorithm/ppo.py (hunk 8 above)
Comment (dongminlee94) [reply maml/algorithm/ppo.py:7]: 신기하네요. 저는 이런 오류가 안떴었는데. 혹시 다시 한번만 더 바꿔서 `make format` 돌려봐주실 수 있으신가요?

File: maml/algorithm/networks.py (hunk 7 above)
Comment (dongminlee94) [reply maml/algorithm/networks.py:8]: 이 부분도 아래의 comment와 같이 한번 더 테스트해봐주시면 감사하겠습니다.

File: maml/configs/experiment_config.yaml (hunk 10 above)
Comment (dongminlee94) [reply maml/configs/experiment_config.yaml:11]: OK. 👍

File: maml/algorithm/meta_learner.py (hunk 1 above)
Comment (Clyde21c) [reply maml/algorithm/meta_learner.py:43]: 네 train_tasks는 전체 갯수로 두고, 샘플링을 한다면 num_sample_tasks는 전체에서 샘플할 개수로 두면 될것 같네요.

File: maml/algorithm/meta_learner.py (hunk 2 above)
Comment (Clyde21c) [reply maml/algorithm/meta_learner.py:69]: 레퍼에선 PPO, TRPO에서 버퍼를 사용 안하는 방식이라 버퍼를 두개 쓰진 않았구요, learn2learn보시면 
https://github.com/learnables/learn2learn/blob/6b741028f0812ba73da58acf7206d93654dde97a/examples/rl/maml_trpo.py#L163
여기에서 inner_loop trajs는 버퍼 없이 바로 학습하고 버리고, 
https://github.com/learnables/learn2learn/blob/6b741028f0812ba73da58acf7206d93654dde97a/examples/rl/maml_trpo.py#L166
outer_loop에서 학습하기위한 각 task별 adapted trajs들은 버퍼에 담아둬요.

사실 저희도 inner_buffer는 데이터를 넣자마자 빼기때문에 없어도 되긴 하다만, 통일성 있게 쓰려다 보니 추가했어요. 이 부분은 동민님 의견 듣고 싶습니다. 
`
trajs = self.sampler.obtain_trajs(inner_policy, max_samples=self.train_samples, max_step=self.max_step)
ppo_loss = self.agent.compute_losses(inner_policy, trajs)
`
learn2learn처럼 한다면 이렇게 바꿔도 될것 같구요.

ray는 좀 독특하게 trajs들을 다뤄서 참고하지 않았습니다.

File: maml/algorithm/meta_learner.py (hunk 6 above)
Comment (Clyde21c) [reply maml/algorithm/meta_learner.py:144]: https://github.com/learnables/learn2learn/blob/6b741028f0812ba73da58acf7206d93654dde97a/examples/rl/maml_trpo.py#L45
넵 learn2learn에서 episode trajectory를 통째로 loss하나 만들어 내는 코드 입니다.

File: maml/algorithm/networks.py (hunk 7 above)
Comment (Clyde21c) [reply maml/algorithm/networks.py:8]: 다시 하니 에러 안뜨네요 허허ㅋㅋㅋㅋ 안 익숙해서 그랬던것 같군요

File: maml/algorithm/ppo.py (hunk 8 above)
Comment (Clyde21c) [reply maml/algorithm/ppo.py:7]: 이것도 다시 하니 에러 안뜨네요 ㅋㅋㅋ 원래 버전으로 통일하겠습니다

File: maml/algorithm/meta_learner.py (hunk 1 above)
Comment (dongminlee94) [reply maml/algorithm/meta_learner.py:43]: 네 그러면 잊어버리지 않게 issue에도 올려주세요. pearl과 rl^2에도 추가해야겠네요.

File: maml/algorithm/meta_learner.py (hunk 2 above)
Comment (dongminlee94) [reply maml/algorithm/meta_learner.py:69]: 음 이 부분은 구두로 말하는게 좋겠네요. 일정 한번 잡으시죠 ㅎㅎ

File: maml/algorithm/meta_learner.py (hunk 6 above)
Comment (dongminlee94) [reply maml/algorithm/meta_learner.py:144]: OK. agreed.

File: maml/algorithm/networks.py (hunk 7 above)
Comment (dongminlee94) [reply maml/algorithm/networks.py:8]: Gooooood

File: maml/algorithm/ppo.py (hunk 8 above)
Comment (dongminlee94) [reply maml/algorithm/ppo.py:7]: Gooooood

**Comment by dongminlee94** [comment]
@Clyde21c  리뷰 중 합의가 이루어진 대화들 (or 결론이 난 대화들)은 `resolve conversation` 해주시면 될 것 같습니다.

File: maml/algorithm/meta_learner.py (hunk 2 above)
Comment (Clyde21c) [reply maml/algorithm/meta_learner.py:69]: 네 그러시죠 ㅎㅎ

**Comment by Clyde21c** [comment]
@dongminlee94 결론 난 대화 모두 resolve conversation 했습니다 :) requirement.txt에 higher추가한것을 포함하여 기타 포맷 통일시킨 commit 푸시해서 추가하였습니다. runnability 체크 부탁드릴게요!

**Comment by dongminlee94** [comment]
> @dongminlee94 결론 난 대화 모두 resolve conversation 했습니다 :) requirement.txt에 higher추가한것을 포함하여 기타 포맷 통일시킨 commit 푸시해서 추가하였습니다. runnability 체크 부탁드릴게요!

Runnability of the code is checked. LGTM

**Review by dongminlee94 — APPROVED** [review]
LGTM
